Add fields index and type to rated documents

MaineC opened the pull request:
@cbuescher this adds index and type information to the document ratings. It also add roundtrip parsing for the RatedDocument XContent part. Would be great if you could have a look. If all is fine and jenkins is happy, please feel free to go ahead and hit merge as soon as tests are through.

cbuescher commented on modules/rank-eval/src/main/java/org/elasticsearch/index/rankeval/RatedDocument.java:
```diff
@@ -56,31 +106,44 @@ public int getRating() {
-                throw new ParsingException(parser.getTokenLocation(), "unexpected token [{}] while parsing rated document",
-                        token);
-            }
+    public static RatedDocument fromXContent(XContentParser parser, RankEvalContext context) throws IOException {
```
Nit: If we don't really need the context in the parser (e.g. for parsing nested queries) and this argument is only here to satisfy ObjectParse.parse(), then this could be simply ParseFieldMatcherSupplier.

cbuescher commented:
@MaineC this looks good, left one small comment. Also it looks to me like the added index/type information isn't used for linking rated docs and search results in the metrics yet. I think it would be good to add that to this PR as well.

MaineC commented:
@cbuescher Update. I added functionality to check index/type information when comparing rated docs to actual hits retrieved. Would be great if you could have a look.

MaineC commented:
@cbuescher merged base in, can you take a look please?

cbuescher commented on modules/rank-eval/src/main/java/org/elasticsearch/index/rankeval/RatedDocument.java:
```diff
@@ -19,36 +19,69 @@
+    private RatedDocumentKey key;
+    private int rating;
+
+    RatedDocument() {}
```
Can we make the key and rating mandatory ctor args if this doesn't create too much overhead? The parser needs to be changed to ConstructingObjectParser then.

cbuescher commented on modules/rank-eval/src/main/java/org/elasticsearch/index/rankeval/RatedDocumentKey.java:
```diff
@@ -0,0 +1,130 @@
+    private String type;
+    private String index;
+
+    public RatedDocumentKey() {}
```
Same here, I think the ctor should contain the mandatory arguments and those should be immutable later. Again, this means for parsing we need ConstructingObjectParser.

cbuescher commented on modules/rank-eval/src/test/java/org/elasticsearch/index/rankeval/DiscountedCumulativeGainAtTests.java:
```diff
@@ -95,12 +97,13 @@ public void testDCGAtSix() throws IOException, InterruptedException, ExecutionEx
+                rated.add(new RatedDocument(new RatedDocumentKey("index", "type", Integer.toString(i)), relevanceRatings[i]));
             }
             hits[i] = new InternalSearchHit(i, Integer.toString(i), new Text("type"), Collections.emptyMap());
+            hits[i].shard(new SearchShardTarget("testnode", new ShardId("index", "uuid", 0)));
```
Just curious what this does, in case I need it in another test again.

MaineC replied:
It adds information to the hit about which index it comes from. Needed since we now check the triplet index/type/doc_id instead of just doc_id

cbuescher commented:
@MaineC left a few small comments, other than that I like the change and the introduction of the RatedDocumentKey object.

MaineC commented:
.oO(Glad github lets you re-open closed pull requests - browser made this side jump in just the right moment for me to hit the wrong button instead of into the textfield I'm typing in right now...)

@cbuescher addressed your comments, thanks for pointing me to ConstructingObjectParser (and thanks to @nik9000 for including really helpful documentation on how to actually use it!)

cbuescher commented:
LGTM, thanks for the clarification
